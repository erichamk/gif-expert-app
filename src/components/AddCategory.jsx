import PropTypes from 'prop-types';
import { useState } from 'react';

export const AddCategory = ({ setCategories, onNewCategory}) => {
	const [inputValue, setInputValue] = useState('');

	const onInputChange = (event) => {
		setInputValue(event.target.value);
	};

	const onSubmit = (event) => {
		// para que no haga refresh de la pagina
		event.preventDefault();
		console.log(inputValue);
		if (inputValue.trim().length <= 1) return;
		// Metodo 1: llamar al hook directamente (no recomendado)
		// como no se tiene referencia al array original:
		// setCategories((categories) => [inputValue, ...categories]);
		// MÃ©todo 2: emitir un evento al componente padre para que llame al hook
		onNewCategory(inputValue.trim());
		setInputValue('');
	};

	return (
		<form onSubmit={onSubmit} aria-label="form">
			<input
				type='text'
				placeholder='Buscar gifs'
				value={inputValue}
				onChange={onInputChange}
			/>
		</form>
	);
};

AddCategory.propTypes = {
	onNewCategory: PropTypes.func.isRequired
}