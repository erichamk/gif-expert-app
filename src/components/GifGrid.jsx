import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {

	// custom hook
	const {images, isLoading} = useFetchGifs(category)

	return (
		<>
			<h3>{category}</h3>
			{
				isLoading && (<h2>Cargando...</h2>)
			}
			<div className='card-grid'>
				{/* atender que => () retorna lo que estÃ¡ en el () pero => {} solo ejecuta */}
				{images.map((image) => (
					<GifItem key={image.id} {...image} />
				))}
			</div>
		</>
	);
};

GifGrid.propTypes = {
	category: PropTypes.string.isRequired
}