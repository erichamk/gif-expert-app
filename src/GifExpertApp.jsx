import { useState } from 'react';
// import { AddCategory } from './components/AddCategory';
// import { GifGrid } from './components/GifGrid';
import { AddCategory, GifGrid } from './components';

export const GifExpertApp = () => {
	const [categories, setCategories] = useState(['SFW']);

	const onAddCategory = (value) => {
		if (categories.includes(value)) {
			return;
		}
		// categories.push() evitar, no va a funcionar y no hay que mutar estados
		setCategories([value, ...categories]);
		// si no se tiene referencia al array original:
		// setCategories(cats => [...cats, "AAA"])
	};

	return (
		<>
			{/* titulo */}
			<h1>GifExpertApp</h1>

			{/* Input */}
			<AddCategory
				setCategories={setCategories}
				onNewCategory={(value) => onAddCategory(value)}
			/>

			{/* listado de gif */}
			{/* <button onClick={onAddCategory}>Agregar</button> */}
			{categories.map((category) => (
				<GifGrid key={category} category={category} />
			))}
		</>
	);
};
