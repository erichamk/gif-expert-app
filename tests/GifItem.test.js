import { GifItem } from '../src/components';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Pruebas en <GifItem />', () => {
	const id = 'id';
	const title = 'cat GIF';
	const url = 'https://www.url.com/';

	test('debe hacer match con el snapshot', () => {
		const { container } = render(
			<GifItem id={id} title={title} url={url} />
		);

		expect(container).toMatchSnapshot();
	});

	test('debe mostrar la imagen con el url y alt indicado', () => {
		render(<GifItem id={id} title={title} url={url} />);
		//screen.debug()
		// expect(screen.getByRole('img').alt).toBe(url);
		// expect(screen.getByRole('img').alt).toBe(title);
		const { src, alt } = screen.getByRole('img');
		expect(src).toBe(url);
		expect(alt).toBe(title);
	});

	test('debe mostrar lel titulo en el componente', () => {
		render(<GifItem id={id} title={title} url={url} />);
		//screen.debug()
        // getByText() solo busca en texto visible, no properties 
		expect(screen.getByText(title)).toBeTruthy;
	});
});
