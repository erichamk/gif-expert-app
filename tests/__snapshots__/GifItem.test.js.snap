// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="cat GIF"
      src="https://www.url.com/"
    />
    <p>
      cat GIF
    </p>
  </div>
</div>
`;
